src/Common/tests/System/Collections/IDictionary.Generic.Tests.cs

```diff
@@ -0,0 +1,834 @@
+            while (dictionary.ContainsKey(missingKey) || missingKey.Equals(default(TKey)))
+                missingKey = TKeyFactory(seed++);
+            return missingKey;
+        }
+
+        /// <summary>
+        /// For a Dictionary, the key comparer is primarily important rather than the KeyValuePair. For this
+        /// reason, we rely only on the KeyComparer methods instead of the GetIComparer methods.
+        /// </summary>
+        public virtual IEqualityComparer<TKey> GetKeyIEqualityComparer()
+        {
+            return EqualityComparer<TKey>.Default;
+        }
+
+        /// <summary>
+        /// For a Dictionary, the key comparer is primarily important rather than the KeyValuePair. For this
+        /// reason, we rely only on the KeyComparer methods instead of the GetIComparer methods.
+        /// </summary>
+        public virtual IComparer<TKey> GetKeyIComparer()
+        {
+            return Comparer<TKey>.Default;
+        }
+
+        /// <summary>
+        /// Class to provide an indirection around a Key comparer. Allows us to use a key comparer as a KeyValuePair comparer
+        /// by only looking at the key of a KeyValuePair.
+        /// </summary>
+        public class KVPComparer : IEqualityComparer<KeyValuePair<TKey, TValue>>, IComparer<KeyValuePair<TKey, TValue>>
+        {
+            private IComparer<TKey> _comparer;
+            private IEqualityComparer<TKey> _equalityComparer;
+
+            public KVPComparer(IComparer<TKey> comparer, IEqualityComparer<TKey> eq)
+            {
+                _comparer = comparer;
+                _equalityComparer = eq;
+            }
+
+            public int Compare(KeyValuePair<TKey, TValue> x, KeyValuePair<TKey, TValue> y)
+            {
+                return _comparer.Compare(x.Key, y.Key);
+            }
+
+            public bool Equals(KeyValuePair<TKey, TValue> x, KeyValuePair<TKey, TValue> y)
+            {
+                return _equalityComparer.Equals(x.Key, y.Key);
+            }
+
+            public int GetHashCode(KeyValuePair<TKey, TValue> obj)
+            {
+                return _equalityComparer.GetHashCode(obj.Key);
+            }
+        }
+
+        #endregion
+
+        #region ICollection<KeyValuePair<TKey, TValue>> Helper Methods
+
+        protected override ICollection<KeyValuePair<TKey, TValue>> GenericICollectionFactory()
+        {
+            return GenericIDictionaryFactory();
+        }
+
+        protected override ICollection<KeyValuePair<TKey, TValue>> GenericICollectionFactory(int count)
+        {
+            return GenericIDictionaryFactory(count);
+        }
+
+        protected override bool DefaultValueAllowed { get { return false; } }
+
+        protected override bool DuplicateValuesAllowed { get { return false; } }
+
+        protected override void AddToCollection(ICollection<KeyValuePair<TKey, TValue>> collection, int numberOfItemsToAdd)
+        {
+            Debug.Assert(!IsReadOnly);
```

Comment:
Should this be Assert.False(IsReadOnly)?  We want this assert even for release builds, right?